[llvm-objdump] Add support for HIP offload bundles

david-salinas opened the pull request:
Utilize the new extensions to the LLVM Offloading API to extend to llvm-objdump to handle dumping fatbin offload bundles generated by HIP.  This extension to llvm-objdump adds the option --offload-fatbin.  Specifying this option will take the input object/executable and extract all offload fatbin bundle entries into distinct code object files with names reflecting the source file name combined with the Bundle Entry ID.  Users can also use the --arch-name option to filter offload fatbin bundle entries by their target triple.

david-salinas committed:
Extend LLVM Offloading API for binary fatbin Bundles

  With the intention to provide a common API for offloading, this
  extension to the existing LLVM Offloading API adds support for
  Binary Fatbin Bundles; moving some support from the Clang offloading
  API.  The intention is to add functionality to LLVM tooling for
  Binary Fatbin Bundles in subsequent commits.

Change-Id: I907fdcbcd0545162a0ce1cf17ebf7c9f3a4dbde6

Commit:
Extend llvm-objdump to support clang offload fat bins

  extend option --offloading

Change-Id: Ibc865f80e30aa1a6e5495ecfe617be68a5e15fcf

MaskRay commented on llvm/test/tools/llvm-objdump/Offloading/fatbin-offloading.test:
```diff
@@ -0,0 +1,80 @@
+// RUN: clang++ -x hip --hip-link --offload-arch=gfx1100 --offload-arch=gfx1101 -o %t %s
```
Due to layering (LLVM_ENABLE_PROJECTS may not include clang), llvm/test tests cannot use clang

david-salinas replied:
I'm actually going to just remove this test, in favour of using the other one, which just uses yaml2obj.

lamb-j commented on llvm/test/tools/llvm-objdump/Offloading/fatbin-offloading.test:
```diff
@@ -0,0 +1,80 @@
+    uint32_t A_ref = 3*i;
+    if( A_h[i] != A_ref )
+    {
+      std::cout<< "Mismatch occured at " << i << ": " << A_h[i] << " != " << A_ref << std::endl;
```
space after cout?

david-salinas replied:
I'm going to remove this test, in favour of the other one which just uses yaml2obj

lamb-j commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -23,14 +24,20 @@
 namespace {
 
+static llvm::TimerGroup
+    ClangOffloadBundlerTimerGroup("Clang Offload Bundler Timer Group",
```
We could rename this "OffloadBundlerTimerGroup".

I've been thinking about the newly seperated bundler APIs as the "Offload Bundler", and the clang-offload-bundler binary tool as the "Clang Offload Bundler"

david-salinas replied:
Good point.  This code is no longer in Clang, so no need for that in the name.

david-salinas replied:
resolved in tlatest patch

jplehr commented:
I wanted to run this through some testing but it seems that the current version does not build.

david-salinas commented:
> I wanted to run this through some testing but it seems that the current version does not build.

@jplehr i've pushed a new patch that resolves the build issue.

jplehr commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -379,3 +622,223 @@ bool object::areTargetsCompatible(const OffloadFile::TargetID &LHS,
+
+llvm::Expected<std::unique_ptr<llvm::MemoryBuffer>>
+CompressedOffloadBundle::decompress(llvm::MemoryBufferRef &Input,
+
```
```suggestion
```

david-salinas replied:
ok

jplehr reviewed (commented):
I left a few minor comments and ran my local testing via the buildbot configs I have. That all seems to work. Thanks.

lamb-j commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,160 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+        str = str.drop_front(OffsetStr.size());
+
+        if (str.consume_front("&size=")) {
+          Size;
```
Getting a warning here:

warning: statement has no effect [-Wunused-value]
  331 |           Size;

david-salinas replied:
left over edit

lamb-j commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,160 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+       << ", ID Length = " << IDLength << ", ID = " << ID;
+  }
+  void dumpURI(raw_ostream &OS, StringRef filePath) {
+    OS << ID.data() << "\tfile:\/\/" << filePath << "#offset=" << Offset
```
Getting a warning here:

 warning: unknown escape sequence: '\/'
  225 |     OS << ID.data() << "\tfile:\/\/" << filePath << "#offset=" << Offset

david-salinas replied:
will correct

david-salinas commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,160 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+  create(MemoryBufferRef, uint64_t SectionOffset, StringRef fileName);
+  Error extractBundle(const ObjectFile &Source);
+
+  Error DumpEntryToCodeObject();
```
ok

david-salinas commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,160 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+    }
+  }
+
+  OffloadBundleFatBin(MemoryBufferRef Source, StringRef file) : FileName(file) {
```
ok

david-salinas commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,160 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+  }
+
+  void PrintEntriesAsURI() {
+    SmallVectorImpl<OffloadBundleEntry>::iterator it = Entries.begin();
```
ok.  Will change the other loop to use Entries.size()

david-salinas commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -167,6 +216,102 @@ Error extractFromArchive(const Archive &Library,
+
+  // read the number of Code Objects (Entries) in the current Bundle.
+  if (auto EC = Reader.readInteger(NumOfEntries)) {
+    printf("OffloadBundleFatBin::ReadEntries .... failed to read number of "
```
left over debug stmt

jh7370 commented on llvm/lib/Object/ObjectFile.cpp:
```diff
@@ -212,7 +212,6 @@ ObjectFile::createObjectFile(StringRef ObjectPath) {
   if (std::error_code EC = FileOrErr.getError())
     return errorCodeToError(EC);
   std::unique_ptr<MemoryBuffer> Buffer = std::move(FileOrErr.get());
-
```
Unrelated change?

david-salinas replied:
yes.

jh7370 reviewed (commented):
Please remember to update the llvm-objdump command guide when functionality (especially a new option) is added/changed.

david-salinas commented:
> Please remember to update the llvm-objdump command guide when functionality (especially a new option) is added/changed.

I will add in next patch

jh7370 reviewed (commented):
Doc changes and other generic stuff look good to me. Others with offloading knowledge will need to review the rest.

jhuber6 reviewed (commented):
Long term I think the offload bundles should go away entirely. If this really needs to exist, then at a minimum these need to share the same option. You can detect the format through the magic bits, no need for the confusion.

yxsamliu commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -49,6 +51,31 @@ enum ImageKind : uint16_t {
   IMG_LAST,
 };
 
+class CompressedOffloadBundle {
```
need to be updated to support v3. and refactor clang to use this header.

david-salinas replied:
Yes.  Though I was thinking I'd update this and refactor clang to use the LLVM Offloading API in another patch.

yxsamliu replied:
I am OK with that

david-salinas commented:
> Long term I think the offload bundles should go away entirely. If this really needs to exist, then at a minimum these need to share the same option. You can detect the format through the magic bits, no need for the confusion.

@yxsamliu any objections to just having the --offloading option, and dropping "--offload-fatbin"?  So, we would attempt to extract/dump Offload Fat Binaries with just --offloading.

yxsamliu commented:
> > Long term I think the offload bundles should go away entirely. If this really needs to exist, then at a minimum these need to share the same option. You can detect the format through the magic bits, no need for the confusion.
> 
> @yxsamliu any objections to just having the --offloading option, and dropping "--offload-fatbin"? So, we would attempt to extract/dump Offload Fat Binaries with just --offloading.

No objection.

david-salinas commented:
I've changed this PR so that the existing --offloading option is used to extract clang fatbin offload bundles, and removed the --offload-fatbin option.  And --arch-name is still used to select fatbin bundles for specific target archs.

jhuber6 commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -218,6 +218,8 @@ OPTIONS
 .. option:: --offloading
 
   Display the content of the LLVM offloading section.
+  Support added for extracting Clang Fat Binaries, which will
+  respect the :option `--arch-name` option.
```
This is documentation, not release notes. It should explain what it does.

jh7370 replied:
+1. You could add a release note to cover this though (which is a separate document).

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,159 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+
+/// Fat binary embedded in object files in clang-offload-bundler format
+class OffloadBundleFatBin {
+
+private:
```
```suggestion
```
Nit. redundant.

david-salinas replied:
ok

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,159 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+    SmallVector<OffloadBundleEntry> found = SmallVector<OffloadBundleEntry>();
+    SmallVectorImpl<OffloadBundleEntry>::iterator it = Entries.begin();
+    for (uint64_t I = 0; I < Entries.size(); I++) {
+      if (it->ID.contains(str)) {
+        found.push_back(*it);
+      }
```
```suggestion
      if (it->ID.contains(str))
        found.push_back(*it);
```

david-salinas replied:
ok

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,159 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+        Entries(SmallVector<OffloadBundleEntry>()) {}
+
+  SmallVector<OffloadBundleEntry> EntryIDContains(StringRef str) {
+    SmallVector<OffloadBundleEntry> found = SmallVector<OffloadBundleEntry>();
```
Lots of LLVM style nits, `clang-tidy` from your LSP should let you automatically fix them. LLVM is `PascalCase` for variables.

david-salinas replied:
ok

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,159 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+
+  SmallVector<OffloadBundleEntry> EntryIDContains(StringRef str) {
+    SmallVector<OffloadBundleEntry> found = SmallVector<OffloadBundleEntry>();
+    SmallVectorImpl<OffloadBundleEntry>::iterator it = Entries.begin();
```
Is this just finding all the entries that contain a string? This would be way easier with `llvm::transform(Entires, std::back_inserter(Found), [](auto &X) { ... })` or something.

david-salinas replied:
yup.  Will change in next patch.

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,159 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+          // Size;
+          str.getAsInteger(10, Size);
+        } else
+          report_fatal_error("Reading 'size' in URI.");
```
API functions like this should never fatally error like this. Return an `llvm::Error` and let the user decide if it's fatal.

david-salinas replied:
👍

david-salinas replied:
👍

jhuber6 replied:
FYi I get a ton of warnings on this `OS << ID.data() << "\tfile:\/\/" << filePath << "#offset=" << Offset`

lamb-j replied:
Likely the same as this one: https://github.com/llvm/llvm-project/pull/114834#discussion_r1925439268

david-salinas replied:
These warnings should be fixed now in latest patch.

jhuber6 commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -96,6 +103,48 @@ Error extractFromObject(const ObjectFile &Obj,
+    else {
+      return Error::success();
+    }
+  } // end of while loop
```
Unnecessary comment

david-salinas replied:
👍

jhuber6 commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -96,6 +103,48 @@ Error extractFromObject(const ObjectFile &Obj,
+      return FatBundleOrErr.takeError();
+    OffloadBundleFatBin &Bundle = **FatBundleOrErr;
+
+    // add current Bundle to list.
```
Commends all need to be properly capitalized.

david-salinas replied:
👍

jhuber6 commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -96,6 +103,48 @@ Error extractFromObject(const ObjectFile &Obj,
+
+    if (nextbundleStart >= 0)
+      Offset += nextbundleStart;
+    else {
+      return Error::success();
```
Why this return? Just let it exit the while loop and return normally.

david-salinas replied:
👍

jhuber6 commented on llvm/lib/Object/OffloadBinary.cpp:
```diff
@@ -167,6 +216,100 @@ Error extractFromArchive(const Archive &Library,
+
+  // Read the Magic String first.
+  StringRef Magic;
+  if (auto EC = Reader.readFixedString(Magic, 24)) {
+    return errorCodeToError(object_error::parse_failed);
+  }
```
```suggestion
  if (auto EC = Reader.readFixedString(Magic, 24))
    return errorCodeToError(object_error::parse_failed);
```
And everywhere else.

david-salinas replied:
👍

jhuber6 reviewed (commented):
Lots of nits, most of the code needs to be looked at again to make sure it follows https://llvm.org/docs/CodingStandards.html. The OffloadBinary format seems surprisingly complicated, I'm surprised this handling wasn't upstream already. Is it mostly being copied from somewhere else? Is that why the style is weird?

jhuber6 commented:
I'm wondering if we should split up the addition of the offloading binary handling. How different are the definitions here from the ones in `clang/Driver/OffloadBundler.h`? Should we just move those to LLVM? I also feel like the definitions should be in a separate header file, since they're distinct from the `OffloadBinary`, totally different binary formats.

yxsamliu commented:
> I'm wondering if we should split up the addition of the offloading binary handling. How different are the definitions here from the ones in `clang/Driver/OffloadBundler.h`? Should we just move those to LLVM? I also feel like the definitions should be in a separate header file, since they're distinct from the `OffloadBinary`, totally different binary formats.

I tend to agree that it is cleaner to have separate header and cpp files for OffloadBundler in LLVM, but we could defer the refactoring of clang code to another PR as David suggested in the reply to one of my previous comments.

jhuber6 commented on llvm/test/tools/llvm-objdump/Offloading/fatbin.test:
```diff
@@ -0,0 +1,845 @@
+## Test that --offload-fatbin works correctly
```
This file is massive, is there no way to generate it like we do with some of the other tests?

jhuber6 replied:
Since we're adding an entirely new object format we probably need some unit tests for those as well. See `llvm/unittests/Object/OffloadingTest.cpp` for the other format's parsing test.

david-salinas replied:
> This file is massive, is there no way to generate it like we do with some of the other tests?

I was able to pare it down a fair amount.  I used an executable built from the "simpleAdd.hip" HIP test as input to obj2yaml, which generated the original larger test file.  I removed a lot of the yaml content that wasn't really needed.  In the end we need an object file with a .hip_fatbin section with at least one bundle, with at least one entry.  We should also probably have a .hip_fatbin section with multiple Bundles, each with their own bundle entries.  But the Bundles have 4k alignment.  If we want to test extracting them and verifying that we get back actual code object data, I'm not sure how we can get around at test that isn't at least this big.  But ... I've also added a unittest in the latest patch.  And it is a lot easier to create the code object data.  Though the unittest I did add needs to test more.  COFF object types and objects with multiple bundles, off the top of my head.

jhuber6 commented on llvm/tools/llvm-objdump/OffloadDump.cpp:
```diff
@@ -48,7 +51,7 @@ static void printBinary(const OffloadBinary &OB, uint64_t Index) {
 
 /// Print the embedded offloading contents of an ObjectFile \p O.
-void llvm::dumpOffloadBinary(const ObjectFile &O) {
+void llvm::dumpOffloadBinary(const ObjectFile &O, std::string ArchName) {
```
```suggestion
void llvm::dumpOffloadBinary(const ObjectFile &O, StringRef ArchName) {
```
Unless we really need to have ownership of this string.

jhuber6 commented on llvm/tools/llvm-objdump/OffloadDump.cpp:
```diff
@@ -64,6 +67,57 @@ void llvm::dumpOffloadBinary(const ObjectFile &O) {
+  // Now filter based on if arch-name is specified
+  SmallVectorImpl<llvm::object::OffloadBundleFatBin>::iterator BundleIter =
+      FoundBundles.begin();
+  for (uint64_t bundle_num = 0; bundle_num < FoundBundles.size();
```
You can probably use `llvm::enumerate` here.

david-salinas replied:
I took a stab at trying to convert these for-loops to enumerate and couldn't quite get it working, but it may be that I don't understand enumerate well enough  :-) .  I should at least change the for loops to range based for loops though.  I'll try again with my next patch.

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -183,11 +210,166 @@ class OffloadFile : public OwningBinary<OffloadBinary> {
+  }
+
+  void printEntriesAsURI() {
+    SmallVectorImpl<OffloadBundleEntry>::iterator It = Entries.begin();
+    for (uint64_t I = 0; I < Entries.size(); I++) {
+      It->dumpURI(outs(), FileName);
+      ++It;
+    }
```
```suggestion
  for (OffloadBundleEntry &Entry : Entries)
    Entry.dumpURI(outs(), Filename);
```
And everywhere else you could trivially replace the for-loop.

jhuber6 commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -218,6 +218,8 @@ OPTIONS
 
   Display the content of the LLVM offloading section.
+  Support added for extracting Clang Fat Binaries, which will
+  respect the :option `--arch-name` option.
```
Still needs to be fixed.

jhuber6 commented on llvm/include/llvm/Object/OffloadBinary.h:
```diff
@@ -49,6 +49,7 @@ enum ImageKind : uint16_t {
   IMG_LAST,
 };
 
+
```
```suggestion
```

jhuber6 commented on llvm/include/llvm/Object/OffloadBundle.h:
```diff
@@ -0,0 +1,213 @@
+
+  Error readEntries(StringRef Section, uint64_t SectionOffset);
+  void dumpEntries() {
+    for (OffloadBundleEntry &Entry : Entries) {
+      Entry.dumpInfo(outs());
+    }
```
```suggestion
    for (OffloadBundleEntry &Entry : Entries)
      Entry.dumpInfo(outs());
```
And below

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,482 @@
+                                       uint64_t SectionOffset) {
+  uint64_t NumOfEntries = 0;
+
+  // get Reader
```
```suggestion
```
Nit, comments should be grammatically correct sentences and helpful.

jhuber6 commented on llvm/tools/llvm-objdump/OffloadDump.cpp:
```diff
@@ -64,6 +69,58 @@ void llvm::dumpOffloadBinary(const ObjectFile &O) {
+      // TODO: If -d was specified, disasseble the Code Object too
+
+      ++FoundIter;
+    } // end of for found_entries loop
```
Useless comments.

jhuber6 reviewed (commented):
Looking much better, just a few nits.

jhuber6 commented on llvm/tools/llvm-objdump/OffloadDump.cpp:
```diff
@@ -64,6 +68,56 @@ void llvm::dumpOffloadBinary(const ObjectFile &O) {
+// if -d is specified, disassemble the Code Object Files
+// if -arch=-name is specified, only dump the Entries that match the target arch
+void llvm::dumpOffloadBundleFatBinary(const ObjectFile &O, StringRef ArchName) {
+  assert((O.isELF() || O.isCOFF()) && "Invalid file type");
```
This should be a warning like it is above, assertions are used for diagnosing invariants and not invalid input. Without an assertions build this check won't exist so someone calling it with a Mach-O file will do something unexpected.

There should also be some tests for these types of warnings / errors. I think all the cases that the existing offload tests cover.

jhuber6 commented on llvm/tools/llvm-objdump/OffloadDump.cpp:
```diff
@@ -64,6 +68,47 @@ void llvm::dumpOffloadBinary(const ObjectFile &O) {
+    reportError(O.getFileName(), "while extracting offload FatBin bundles: " +
+                                     toString(std::move(Err)));
+
+  for (const auto &[bundle_num, Bundle] : llvm::enumerate(FoundBundles)) {
```
```suggestion
  for (const auto &[BundleNum, Bundle] : llvm::enumerate(FoundBundles)) {
```

jhuber6 commented on llvm/include/llvm/Object/OffloadBundle.h:
```diff
@@ -0,0 +1,212 @@
+//===--- OffloadBundle.h - Utilities for handling offloading code  -*- C++
+//-*-===//
```
Fix

jhuber6 commented on llvm/include/llvm/Object/OffloadBundle.h:
```diff
@@ -0,0 +1,212 @@
+    int64_t O = 0;
+    int64_t S = 0;
+
+    if (Str.consume_front("file://")) {
```
Not a fan of the nesting here, can we do instead `if (!Str.consume_front) error`?

jhuber6 commented on llvm/docs/CommandGuide/llvm-objdump.rst:
```diff
@@ -217,7 +217,7 @@ OPTIONS
 .. option:: --offloading
 
-  Display the content of the LLVM offloading section.
+  Display the content of the LLVM offloading sections including Clang Fat Binaries.
```
```suggestion
  Display the content of the LLVM offloading sections and HIP offload bundles.
```

jhuber6 commented on llvm/include/llvm/Object/OffloadBundle.h:
```diff
@@ -0,0 +1,212 @@
+//===--- OffloadBundle.h - Utilities for handling offloading code -*- C++
+//-*-===//
```
Fix

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,477 @@
+//===- OffloadBundle.cpp - Utilities for handling offloading code  -*- C++
```
Fix

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,477 @@
+
+  // Read the Magic String first.
+  StringRef Magic;
+  if (auto EC = Reader.readFixedString(Magic, 24)) {
+    return errorCodeToError(object_error::parse_failed);
+  }
```
```suggestion
  if (auto EC = Reader.readFixedString(Magic, 24))
    return errorCodeToError(object_error::parse_failed);
```

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,477 @@
+    if (auto EC = Reader.readFixedString(EntryID, EntryIDSize))
+      return errorCodeToError(object_error::parse_failed);
+
+    // create a Bundle Entry object:
```
```suggestion
    // Create  a Bundle Entry object.
```

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,477 @@
+  // This will extract all entries in the Bundle
+  for (OffloadBundleEntry &Entry : Entries) {
+
+    if (Entry.Size > 0) {
```
```suggestion
    if (Entry.Size == 0)
      continue
```
nit. prefer early exits

jhuber6 reviewed (commented):
Mostly nits, please fix the file headers.

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,475 @@
+      return errorCodeToError(object_error::parse_failed);
+
+    // Create a Bundle Entry object:
+    auto Entry = new OffloadBundleEntry(EntryOffset + SectionOffset, EntrySize,
+                                        EntryIDSize, EntryID);
```
Sorry this one's a little late, but any reason we can't use `std::make_unique` here?

jhuber6 commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,478 @@
+    if (auto EC = Reader.readFixedString(EntryID, EntryIDSize))
+      return errorCodeToError(object_error::parse_failed);
+
+    // Create a Bundle Entry object:
+    //    auto Entry = new OffloadBundleEntry(EntryOffset + SectionOffset,
+    //    EntrySize,
+    //                                        EntryIDSize, EntryID);
```
```suggestion
```
Nit

jhuber6 reviewed (approved):
Seems fine for now.

david-salinas commented:
I've verified this patch with Top of Tree in branch "main".

jhuber6 commented:
```
FAILED: libc/src/stdlib/CMakeFiles/libc.src.stdlib.memalignment.dir/memalignment.cpp.o 
/home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.build/./bin/clang++ --target=amdgcn-amd-amdhsa -DLIBC_NAMESPACE=__llvm_libc_21_0_0_git -D__LIBC_USE_FLOAT16_CONVERSION -I/home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.src/libc -isystem /home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.build/include/amdgcn-amd-amdhsa -O3 -DNDEBUG --target=amdgcn-amd-amdhsa -DLIBC_QSORT_IMPL=LIBC_QSORT_QUICK_SORT -DLIBC_ADD_NULL_CHECKS "-DLIBC_MATH=(LIBC_MATH_SKIP_ACCURATE_PASS | LIBC_MATH_SMALL_TABLES | LIBC_MATH_NO_ERRNO | LIBC_MATH_NO_EXCEPT)" -fpie -DLIBC_FULL_BUILD -nostdlibinc -ffixed-point -fno-exceptions -fno-lax-vector-conversions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -ftrivial-auto-var-init=pattern -fno-omit-frame-pointer -Wall -Wextra -Werror -Wconversion -Wno-sign-conversion -Wdeprecated -Wno-c99-extensions -Wno-gnu-imaginary-constant -Wno-pedantic -Wimplicit-fallthrough -Wwrite-strings -Wextra-semi -Wnewline-eof -Wnonportable-system-include-path -Wstrict-prototypes -Wthread-safety -Wglobal-constructors -nogpulib -fvisibility=hidden -fconvergent-functions -flto -Wno-multi-gpu -Xclang -mcode-object-version=none -DLIBC_COPT_PUBLIC_PACKAGING -UNDEBUG -MD -MT libc/src/stdlib/CMakeFiles/libc.src.stdlib.memalignment.dir/memalignment.cpp.o -MF libc/src/stdlib/CMakeFiles/libc.src.stdlib.memalignment.dir/memalignment.cpp.o.d -o libc/src/stdlib/CMakeFiles/libc.src.stdlib.memalignment.dir/memalignment.cpp.o -c /home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.src/libc/src/stdlib/memalignment.cpp
/home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.src/libc/src/stdlib/memalignment.cpp:20:3: error: unknown type name 'uintptr_t'
   20 |   uintptr_t addr = reinterpret_cast<uintptr_t>(p);
      |   ^
/home/botworker/builds/openmp-offload-amdgpu-runtime-2/llvm.src/libc/src/stdlib/memalignment.cpp:20:37: error: unknown type name 'uintptr_t'
   20 |   uintptr_t addr = reinterpret_cast<uintptr_t>(p);
      |                                     ^
2 errors generated.
```
Looks unrelated, no clue why this is erroring. I'll look into it.

alinas commented on llvm/lib/Object/OffloadBundle.cpp:
```diff
@@ -0,0 +1,478 @@
+        SectionOffset = ELFSectionRef(Sec).getOffset();
+      } else if (Obj.isCOFF()) {
+        if (const COFFObjectFile *COFFObj = dyn_cast<COFFObjectFile>(&Obj))
+          const coff_section *CoffSection = COFFObj->getCOFFSection(Sec);
```
This is unused, and is triggering "error: unused variable 'CoffSection' [-Werror,-Wunused-variable]".
If the plan is to extend this soon, please use:
```
(void) CoffSection
```
Otherwise, please remove branch and re-add when adding functionality.

Kewen12 commented:
Hi David, could you please revert the PR while investigating? It breaks our buildbot and blocks downstream merge. Thanks!

kazutakahirata commented:
@david-salinas @Kewen12 I've reverted the PR.

kazutakahirata commented:
@david-salinas If you happen to have clang available as the host compiler, you can run `cmake` with:

```
-DLLVM_ENABLE_WERROR=On \
-DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
```

This way, you can easily catch warnings without relying on build bots.  The LLVM project usually builds without warnings as long as you use clang as the host compiler.
